File: C++/design-neighbor-sum-service.cpp
// Time:  ctor:        O(n^2)
//        adjacentSum: O(1)
//        diagonalSum: O(1)
// Space: O(n^2)

// hash table
class neighborSum {
public:
    neighborSum(vector<vector<int>>& grid)
        : grid_(grid)
        , lookup_(size(grid) * size(grid[0]))
    {
        for (int i = 0; i < size(grid); ++i) {
            for (int j = 0; j < size(grid[0]); ++j) {
                lookup_[grid[i][j]] = {i, j};
            }
        }
    }
    
    int adjacentSum(int value) {
        static const vector<pair<int, int>> ADJACENTS{{1, 0}, {0, 1}, {-1, 0}, {0, -1}};
    
        return sum(value, ADJACENTS);
    }
    
    int diagonalSum(int value) {
        static const vector<pair<int, int>> DIAGONALS{{1, 1}, {1, -1}, {-1, 1}, {-1, -1}};

        return sum(value, DIAGONALS);
    }

private:
    int sum(int value, const auto& directions) {
        const auto& [i, j] = lookup_[value];
        int total = 0;
        for (const auto& [di, dj] : directions) {
            const int ni = i + di, nj = j + dj;
            if (!(0 <= ni && ni < size(grid_) && 0 <= nj && nj < size(grid_[0]))) {
                continue;
            }
            total += grid_[ni][nj];
        }
        return total;
    }

    vector<vector<int>> grid_;
    vector<pair<int, int>> lookup_;
};

File: Python/design-neighbor-sum-service.py
# Time:  ctor:        O(n^2)
#        adjacentSum: O(1)
#.       diagonalSum: O(1)
# Space: O(n^2)

# hash table
class neighborSum(object):
    ADJACENTS = ((1, 0), (0, 1), (-1, 0), (0, -1))
    DIAGONALS = ((1, 1), (1, -1), (-1, 1), (-1, -1))


    def __init__(self, grid):
        """
        :type grid: List[List[int]]
        """
        self.__grid = grid
        self.__lookup = [None]*(len(grid)*len(grid[0]))
        for i in xrange(len(grid)):
            for j in xrange(len(grid[0])):
                self.__lookup[grid[i][j]] = (i, j)


    def adjacentSum(self, value):
        """
        :type value: int
        :rtype: int
        """
        return self.__sum(value, neighborSum.ADJACENTS)
    

    def diagonalSum(self, value):
        """
        :type value: int
        :rtype: int
        """
        return self.__sum(value, neighborSum.DIAGONALS)


    def __sum(self, value, directions):
        i, j = self.__lookup[value]
        total = 0
        for di, dj in directions:
            ni, nj = i+di, j+dj
            if not (0 <= ni < len(self.__grid) and 0 <= nj < len(self.__grid[0])):
                continue
            total += self.__grid[ni][nj]
        return total

